core/src/main/java/org/apache/iceberg/util/SnapshotUtil.java

```diff
@@ -144,4 +146,72 @@ public static Snapshot snapshotAfter(Table table, long snapshotId) {
     throw new IllegalStateException(
         String.format("Cannot find snapshot after %s: not an ancestor of table's current snapshot", snapshotId));
   }
+
+  /**
+   * Returns the ID of the most recent snapshot for the table as of the timestamp.
+   *
+   * @param table a {@link Table}
+   * @param timestampMillis the timestamp in millis since the epoch
+   * @return the snapshot ID
+   * @throws IllegalArgumentException when no snapshot is found in the table
+   * older than the timestamp
+   */
+  public static long snapshotIdAsOfTime(Table table, long timestampMillis) {
+    Long snapshotId = null;
+    for (HistoryEntry logEntry : table.history()) {
+      if (logEntry.timestampMillis() <= timestampMillis) {
+        snapshotId = logEntry.snapshotId();
+      }
+    }
+    Preconditions.checkArgument(snapshotId != null,
+        "Cannot find a snapshot older than %s", timestampMillis);
+    return snapshotId;
+  }
+
+  /**
+   * Returns the schema of the table for the specified snapshot.
+   *
+   * @param table a {@link Table}
+   * @param snapshotId the ID of the snapshot
+   * @return the schema
+   */
+  public static Schema schemaFor(Table table, long snapshotId) {
+    Snapshot snapshot = table.snapshot(snapshotId);
+    Preconditions.checkArgument(snapshot != null, "Cannot find snapshot %s", snapshotId);
+    Integer schemaId = snapshot.schemaId();
+    // schemaId could be null, if snapshot was created before Iceberg added schema id to snapshot
+    if (schemaId != null) {
+      return table.schemas().get(schemaId);
```

Comment:
How about `getOrDefault(schemaId, table.schema())` just to be safe?

Reply:
shall we allow the current table.schema as the default? IMO, the current behavior is actually confusing, especially when the schema is not compatible, for example 

first snapshot
```
+-----+
|value|
+-----+
|    1|
|    2|
|    3|
|    4|
+-----+
```
overwrite with

```
+---------+
|new_value|
+---------+
|        2|
|        3|
|        4|
|        5|
+---------+
```

time travel to the first one

```
+---------+
|new_value|
+---------+
|        0|
|        0|
|        0|
|        0|
+---------+

```

Reply:
or we should add some configuration entry for users to explicitly allow time traveling to old data but use the latest schema? (IMHO, better than falling into the trouble as I described above)

Reply:
IMO, we should not silently return `table.schema()` in this case; i.e., I don't think we should use `getOrDefault(schemaId, table.schema())`. This is the case where the snapshot does have a `schemaId`, so the schema should be found in `schemas()`. If it isn't, there was some internal inconsistency or corrupt metadata.

We only fall back to returning `table.schema()` as an interim measure (indicated by the TODO) in the following, until the recovery is implemented.

Reply:
@wypoon, that's a fair point. In that case, we should make sure we throw a good error message. Probably use `IllegalStateException`?

Reply:
Yes, I'll throw an `IllegalStateException` if the schema is not found then.